### ISSUE TITLE
[Proposal] Add Hooks and remove After: because don't exist in Gherkin syntax

### ISSUE THREAD

**Issue by lsaudon** [olexale/bdd_widget_test#52]
I think it's possible to add the notion of hooks to the package.
Also, `After:` doesn't exist in Gherkin syntax.

**Suggestion**

Generate a Hooks class like steps.

hooks.dart
```dart
class Hooks {
  static Function before = () {
    // Add logic for before
  };
  static Function beforeStep = () {
    // Add logic for beforeStep
  };
  static Function afterStep = () {
    // Add logic for afterStep
  };
  static Function after = () {
    // Add logic for after
  };
}
```

sample_test.dart
```dart
void main() {
  group('''Counter''', () {
    setUp(() {
      Hooks.before.call();
    });

    tearDown(() {
      Hooks.after.call();
    });

    testWidgets('''Scenario One''', (tester) async {
        Hooks.beforeStep.call();
        await iSeeText(tester, '0');
        Hooks.afterStep.call();
        Hooks.beforeStep.call();
        await iSeeText(tester, '1');
        Hooks.afterStep.call();
    });
  });
}
```

If we can do this.
`After:` has no reason to exist.

**Comment by olexale** [comment]
I'd not make breaking changes if they don't improve someone's life. Let's not remove the `After` keyword, as some users asked for it.

Overall, `Before` and `After` were supposed to replace hooks for most users. I took these names from [here](https://cucumber.io/docs/cucumber/api/?lang=java#hooks) but changed the syntax, i.e. instead of annotating functions in the implementation, you add steps right to scenarios. A horrible thing from the dogmatic BDD point of view, and it violates the `Only use a Before hook for low-level logic such as starting a browser or deleting data from a database.` recommendation by design. However, from what I observed, it significantly simplifies the usage of these hooks and makes them easy to understand for users who don't intend to follow the BDD as it is described in a book.

I see how the proposed "proper hooks" can help with #46, so I was already thinking about something similar. Do you have any other potential use cases in mind?

The proposed functionality looks good (except for the `After` keyword removal, let's not break it for the existing users, unless you see a reason why we need to do this). 
I have a few more questions:
1. How can I specify the hook? Will it always be something like a `test/hook.dart` file with a `Hook` class in it? Or a class should be specified in the config?
2. Would it be possible to chain hooks? If yes — how?
3. Do we want to have conditional hooks that, for example, would be enabled for specific tags only?

**Comment by lsaudon** [comment]
I want the `*.feature` files we create to be compatible with `cucumber` or `specflow`.
That's why I suggested removing `After:`, but it's better to leave it in so as not to have breaking changes with a loss of functionality.

For hook, the package creates skeleton and developers code in the before and after methods.
It's the simplest way I've found (I think it could be improved with a better brain).

```dart
class Hooks {
  static Function before = () {
    otherPart.doStuff(); // Line write by developer
    clearData; // Line write by developer
  };
  ...
}
```

If we need conditional hooks, we just need to change Function like this:

```dart
typedef HooksFunction = void Function(List<String> tags);

class Hooks {
  static HooksFunction get before => (tags) {
    if(tags.contains('important')){
      // do stuff
    }
  };
  static HooksFunction get beforeStep => (tags) {};
  static HooksFunction get afterStep => (tags) {};
  static HooksFunction get after => (tags) {};
}
```

Pros:
- Respect Gherkin syntax and philosophy

Cons:
- Add hooks everywhere, even when not in use (maybe enable by .yml)

**Comment by lsaudon** [comment]
I like the idea and if the person has not entered `hook_import_line` and `hook_class`.
We don't generate the `before`, `after`...
Is that it?

**Comment by olexale** [comment]
Yes, correct. If the hook is not set, we don't generate anything hook-related, so that for users who don't need this functionality (I'd expect the majority of users) widget tests will stay simple.

**Comment by daniel-deboeverie-lemon** [comment]
Since we need this for our own project, we have been looking into how to implement this.

Working with a contract or with a static class has some issue with it. First of all is that we would need to give the import as a string in build.yaml. This is counterintuitive and gives the issue that you can't do package imports if you don't put it in the lib directory, which would be a strange place to put testing steps.

Secondly it requires the developers to read the read.me and do all the steps defined there; edit the build.yaml, create the hooks and they themselves need to figure out and test what works and doesn't for the import.

Thus I propose the following; add a flag in the build.yaml for the use of these hooks and if it is true, generate them as if you would for the step files. Later this could then also be expanded on which of the hooks we want to generate, optionally as separate files.

**Comment by olexale** [comment]
Hey Daniel,

> Since we need this for our own project, we have been looking into how to implement this.

Would you please describe your case? Why do you need hooks?

**Comment by daniel-deboeverie-lemon** [comment]
We wish to add screenshots after each (failed) test. We would do it with the @after hook you have already provided, but we wish to follow the gherkin conventions for the feature files, since they will not only be used in dart.

**Comment by daniel-deboeverie-lemon** [comment]
I am more than willing to write these changes myself and create a PR for you to review.

**Comment by olexale** [comment]
This functionality is available in [1.6.4](https://pub.dev/packages/bdd_widget_test/versions/1.6.4). Thanks @lsaudon for the proposal and @daniel-deboeverie-lemon for the implementation!
